src/State/Processor/RespondProcessor.php

```diff
@@ -75,11 +79,14 @@ public function process(mixed $data, Operation $operation, array $uriVariables =
 
         if ($hasData = ($this->resourceClassResolver && $originalData && \is_object($originalData) && $this->resourceClassResolver->isResourceClass($this->getObjectClass($originalData))) && $this->iriConverter) {
             if (
-                ($operation->getExtraProperties()['is_alternate_resource_metadata'] ?? false)
-                && 301 === $operation->getStatus()
+                300 <= $status && $status < 400
+                && (($operation->getExtraProperties()['is_alternate_resource_metadata'] ?? false) || ($operation->getExtraProperties()['canonical_uri_template'] ?? null))
             ) {
-                $status = 301;
-                $headers['Location'] = $this->iriConverter->getIriFromResource($originalData, UrlGeneratorInterface::ABS_PATH, $operation);
+                $canonicalOperation = $this->resourceMetadataCollectionFactory
+                    ->create($this->getObjectClass($originalData))
```

Comment:
```suggestion
                    ->create($operation->getClass())
```

Although this will not be necessary with the OperationMetadataFactory